DIPs/DIP1xxx-ES.md

```diff
@@ -0,0 +1,98 @@
+# Add an `in` operator for arrays
+
+| Field           | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            | 1013                                                            |
+| Review Count:   | 0 (edited by DIP Manager)                                       |
+| Author:         | Elijah Stone <elronnd@elronnd.net>                              |
+| Implementation: | None yet (TODO)                                                 |
+| Status:         | Draft  |
+
+## Abstract
+
+Asociative arrays and redblacktrees have an `in` operator which will check if a
+variable is present within those data structures.  This operator is also
+present in python for lists, with this syntax:
+
+    lis = [1, 2, 3, 7]
+    3 in lis # True
+    4 in lis # False
+
+
+
+### Reference
+
+The `in` operator was [recently added](https://github.com/dlang/phobos/pull/5629) to `std.range.iota`, showing support for
+useage of the operator in general.
+
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Acknowledgements](#acknowledgements)
+* [Reviews](#reviews)
+
+## Rationale
+
+Although the effects of the proposed `in` operator can be replicated using
+[std.algorithm.searching.canFind](https://dlang.org/library/std/algorithm/searching/can_find.html), the process of importing and calling it is
+cumbersome, and there is a language inconsistency in that AAs and redblacktrees
+have an `in` operator but arrays do not.  Having a simple, quick way of
+checking if an array contains an object in a way that is consistent with the
+rest of the language should be essential.  In addition, the lack of such an
+operator could be jarring to newcomers from such languages as python.  Previous
+criticism of said operator has been that there should not be built-in language
+features with worse than O(1) complexity, but AAs already have the `in` operator,
+which provides worst-case O(log(n)) complexity; and, besides, people who
+actually care about the complexity of their code will know not to use the `in`
+operator, and those who want operators of higher complexity will have an easier
+time using them.
+
+
+## Description
+
+The `in` operator for arrays will take a needle, of type `T`, and a haystack,
+of type `T[]`.  If any of the elements of the haystack are equal to the needle,
+then it will return true.  Otherwise, it will return false.  No new grammar
```

Comment:
I think the right signature would be:
```D
inout(T)* opIn_impl(T)(inout(T)[] haystack, auto ref const(T) needle)
if (__traits(compiles, { bool ok = haystack[0] == needle; }))
```

This is would be consistent with [testing membership in AAs](https://dlang.org/spec/hash-map.html#testing_membership) and will allow easy use of the found element, if any.

Example implementation: https://run.dlang.io/is/elciqv

Reply:
As I mentioned above, returning a pointer is a bad idea since array values aren't guaranteed to be unique, so if I do 5 in [5, 6, 5, 4], I get a pointer only to the  first one.  A solution could be to return an array of pointers or (I prefer this) an array of indices.  Both of those have the advantage that arrays convert implicitly to bool (which might be a bad thing but that's a separate discussion), meaning that if (5 in arr) will still work.